onnxmltools/convert/common/_topology.py

```diff
@@ -700,6 +717,10 @@ def convert_topology(topology, model_name, doc_string, targeted_onnx):
         i += 1
 
     # Add extra information
+    add_metadata_props(onnx_model, metadata_props)
+    for image in image_inputs:
+        set_denotation(onnx_model, image, 'IMAGE',
```

Comment:
This may be true to Core ML. If things comes from Keras, it's not necessary because Keras can have `channels_last=True` where `DATA_CHANNEL` should be moved to the tail of the denotation list. To resolve this problem, we need to add a denotation list into our `[TensorType](https://github.com/onnx/onnxmltools/blob/f8743958da9594488c29fdffb501f53f2f5ecc5d/onnxmltools/convert/common/data_types.py#L56)` and maintain it via the parser.

Reply:
Yes, something like you are describing was in our initial proposal ([see the first commit in this PR](https://github.com/onnx/onnxmltools/pull/115/commits/e5bff90ec457078041f7b0cc9b4edfeb8bcd1650#diff-8ddffe74cd10c33d49c52adee4f9e22b)), but it was suggested that we don't clutter `TensorType` with fields that are currently CoreML specific. I believe our previous approach (the one you are suggesting) is better, should I go back to that?

Reply:
I'll add denotation and channel_denotations parameters to TensorType

Reply:
You will probably need to propagate such information throughout the shape inference stage. This is necessary for some edge cases (our compiling procedure removes identity by replacing its output with its input. If we don't copy metadata from the input to the output in shape inference stage, we will lose it).